realm/realm-library/src/main/java/io/realm/internal/ColumnInfo.java

```diff
@@ -16,60 +16,197 @@
-     * @return shallow copy.
+     * @param src The source for the copy.  This instance will be an exact copy of {@code src} after return.
+     * {@code src} must not be {@code null}.
+     * @throws IllegalArgumentException if {@code other} has different class than this.
      */
+    public final void copyFrom(ColumnInfo src) {
+        if (!mutable) {
+            throw new UnsupportedOperationException("Attempt to modify immutable cache");
+        }
+        if (null == src) {
+            throw new NullPointerException("Attempt to copy null ColumnInfo");
+        }
+
+        indicesMap.clear();
+        indicesMap.putAll(src.indicesMap);  // ColumnDetails are immutable: no need to copy
+        copy(src, this);
+    }
+
     @Override
-    public ColumnInfo clone() {
-        try {
-            return (ColumnInfo) super.clone();
-        } catch (CloneNotSupportedException e) {
-            throw new RuntimeException(e);
+    public String toString() {
+        StringBuilder buf = new StringBuilder("ColumnInfo[");
+        buf.append(mutable).append(",");
+        if (indicesMap != null) {
+            boolean commaNeeded = false;
+            for (Map.Entry<String, ColumnDetails> entry : indicesMap.entrySet()) {
+                if (commaNeeded) { buf.append(","); }
+                buf.append(entry.getKey()).append("->").append(entry.getValue());
+                commaNeeded = true;
+            }
+        }
+        return buf.append("]").toString();
+    }
+
+    /**
+     * Create a new object that is an exact copy of {@code src}.
+     *
+     * @param mutable false to make an immutable copy.
+     */
+    protected abstract ColumnInfo copy(boolean mutable);
+
+    /**
+     * Make {@code dst} into an exact copy of {@code src}.
+     * Intended for use only by subclasses.
+     *
+     * @param src The source for the copy
+     * @param dst The destination of the copy.  Will be an exact copy of src after return.
+     */
+    protected abstract void copy(ColumnInfo src, ColumnInfo dst);
+
+    /**
+     * Add a new column to the indexMap.
+     * <b>For use only by subclasses!</b>.
+     * Must be called from within constructor, to keep effectively-final contract.
+     *
+     * @param realm The shared realm.
+     * @param table The table to search for the column.
+     * @param columnName The name of the column whose index is sought.
+     * @param columnType Type RealmType of the column.
+     * @return the index of the column in the table
+     */
+    @SuppressWarnings("unused")
+    protected final long addColumnDetails(SharedRealm realm, Table table, String columnName, RealmFieldType columnType) {
+        long columnIndex = table.getColumnIndex(columnName);
+        if (columnIndex < 0) {
+            throw new RealmMigrationNeededException(
+                    realm.getPath(),
+                    "Field '" + columnName + "' not found for type " + table.getClassName());
+        }
+
+        // expensive and unnecessary?
```

Comment:
Yes, but I don't believe we need to do it here.  The check happens in Validation, I think.